### PR TITLE
fix(topology): Fix handling of component errors

### PR DESCRIPTION
**Pull request by jszwedko**
In https://github.com/timberio/vector/pull/6238 we inadvertently stopped
handling errors returned by components, only handling panics.

This replaces the "result flattening" code to handle both:

* Panics returned as errors from catch_unwind
* Errors returned from the futures for source / sink components

This uncomments the tests I had commented due to it causing CI to hang
in https://github.com/timberio/vector/pull/6298

Fixes https://github.com/timberio/vector/issues/6299

Signed-off-by: Jesse Szwedko <jesse@szwedko.me>

### CODE REVIEW

**Comment by jszwedko** [comment]
Test failures are #6301

File: src/topology/mod.rs [hunk 1]
```diff
@@ -727,13 +727,18 @@ async fn handle_errors(
...
+        .map_err(|_| ())
+        .and_then(|res| res)
+    {
         Err(_) => {
             error!("An error occurred that vector couldn't handle.");
             let _ = abort_tx.send(());
             Err(())
         }
-        Ok(result) => result,
+        Ok(result) => Ok(result),
```
Comment (bruceg) [inline src/topology/mod.rs:741]: This looks like `.map_err`

**Review by bruceg — APPROVED** [review]
LGTM

File: src/topology/mod.rs (hunk 1 above)
Comment (jszwedko) [reply src/topology/mod.rs:741]: True, `map_err` could be used here. I always feel weird using `map*` methods purely to produce a side-effect (in this case logging and sending on the abort channel). I wish there was a `tap_err` method or similar that would just give me a reference, but not mutate.

Comment (jszwedko) [reply src/topology/mod.rs:741]: I guess someone made a crate for that: https://docs.rs/tap/1.0.0/tap/tap/trait.TapFallible.html

**Comment by jszwedko** [comment]
Mac failures are the flake-y watcher tests #5000

**Pull request merged**